File: src/com/train/algorithm/tree/implementInC++/MostFrequentSubtreeSum.cpp
//
// Created by 罗斯 on 8/11/2022.
//
#include <vector>
#include <unordered_map>

using namespace std;
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    vector<int> findFrequentTreeSum(TreeNode* root) {
        unordered_map<int, int> freq;
        int max_freq = -1;
        vector<int> res;
        (void)treeSum(root, freq, max_freq, res);
        return res;
    }
private:
    int treeSum(TreeNode* root, unordered_map<int, int>& freq, int& max_freq, vector<int>& res){
        if (!root) return 0;
        int sum = root->val + treeSum(root->left, freq, max_freq, res) + treeSum(root->right, freq, max_freq, res);
        int f = ++freq[sum];
        if (f > max_freq){
            max_freq = f;
            res.clear();
        }
        if (f == max_freq){
            res.push_back(sum);
        }
        return sum;
    }
};

File: src/com/train/algorithm/tree/implementInPython/MostFrequentSubtreeSum.py
import collections


class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution(object):
    def findFrequentTreeSum(self, root):
        if not root: return []
        D = collections.defaultdict(int)
        def cal_sum(node):
            if not node: return 0
            rv = node.val + cal_sum(node.left) + cal_sum(node.right)
            D[rv] += 1
            return rv
        cal_sum(root)
        mx = max(D.values())
        return [k for k,v in D.items() if v==mx]